### ISSUE TITLE
GameGuru MAX - Request - Shaders!

### ISSUE THREAD

**Issue by ArgentArts** [TheGameCreators/GameGuruRepo#2948]
I would like to request some way to create, use, import SHADERS in GameGuru MAX. Shaders, I think, are vital to creating modern games and have been for years and years. Here's an OLD example from Half-Life 2:

![glass](https://user-images.githubusercontent.com/36735507/164948293-55232099-1fd3-4a96-891b-23c3004f2f46.png)

I think the ability to create and use shaders are going to be essential.

**Comment by 42Pix3ls** [comment]
I agree, shaders are what take a good-looking game and make it great. You can really hone in your look with shaders and honestly the official release should support them.

**Comment by mav3r1ck198-1** [comment]
Access to existing shaders or the ability to create our own, no matter how challenging, will be critical moving forward to create visually appealing and "modernish" games. For example, the scrolling effect from classic would allow for lava flows or water not attached to the existing water plane to be created (think waterfalls/rivers down mountains etc) This would help to breathe life into the worlds that are being created and add variation to everyone's maps

**Comment by MonkeyFrogStudio** [comment]
> For example, the scrolling effect from classic would allow for lava flows or water

Not only that, but even something as simple as the belt on a conveyer. You really don't want to animated the actual belt, but the material on it to simulate the movement of that belt.

**Comment by mav3r1ck198-1** [comment]
One for you as well, wicked has a custom "hologram" shader that could be useful for your modular kit

**Comment by 42Pix3ls** [comment]
Am I wrong but can't you use shaders to simulate wind in trees and bushes etc?

**Comment by fearlesswee** [comment]
+Support for this, shaders should absolutely be more supported in MAX for those users with the right know-how! I absolutely think having a proper glass shader with refraction and such is essential, because atm anything made of glass in MAX looks pretty strange. (And I think Wicked already has a glass shader/material by default, just a matter of "flicking it on" and giving us the ability to assign shaders to an entity/mesh!)

**Comment by LeeBamberTGC** [comment]
NOTE TO ALL: Does anyone know how to write 'Wicked Shaders' for integration into MAX?  I ask on the groups that I add support, and then no-one actually writes any custom shaders to include :)

**Comment by MonkeyFrogStudio** [comment]
@LeeBamberTGC I don't, because I am not a programmer. But I know that Necrym59, who has written several Behaviors for you and for all of us in general, has been chomping at the bit for some form of shaders. I think he even spent a little time looking at how they are done in Wicked (and what they were based off of). I know that I would most likely even commission someone to write a shader or two for me for my own project. Is that good enough?

**Comment by MonkeyFrogStudio** [comment]
@mav3r1ck198-1 Heck, knowing Graphix, he might write a Wicked Shader editor for us non-programmers to use. ;)

**Comment by LeeBamberTGC** [comment]
@mav3r1ck198-1 @MonkeyFrogStudio Can you ask them to chime in on this thread as I think it is worth adding support if they want to try making a custom shader for MAX. Opens many doors if we have the expertise in the community :)

**Comment by MonkeyFrogStudio** [comment]
@LeeBamberTGC I'd already contacted Necrym59, but he's in Australia and it's late there. I am sure we will hear from him in due time. ;)

**Comment by Necrym59** [comment]
Hi all, I have already looked at the complexity of shaders in Wicked/Max and have come to the conclusion that they are not presently easy to implement. I could be wrong and may revisit it in more depth as I do see a massive need for some though, such as the glass or distortion/deformation and effect shaders above amongst other effects that cant be emulated via other means not yet or could be exposed in MAX, it would  be good in my opinion to get some shaders created based on a community vote list rather than try to create a inbuilt shader creator for the moment as that is a pretty big undertaking for a limited use. 

Perhaps a tool as a dlc, like the particle editor may be the way to go, I would support that idea as it allows a big amount of MAX users to contribute shaders and open up more possibilities, not to mention a possible income stream from its sale, it all depends on the number crunching in my opinion.

A lot of effects people are after can be emulated of course using other techniques such as video/animated textured objects being the most major not everything has to be a shader. Personally i would want to see video and animated textures, just this alone being played on a transparent plane in front of the player camera can open up easier shader like effects for users and as mentioned tv, screens, converor belts almost any object can come alive with animated textures or video.

**Comment by MonkeyFrogStudio** [comment]
> rather than try to create a inbuilt shader creator 

Who said anything about TGC creating an inbuilt shader creator or editor? My OP was about opening up or adding Wicked Shader ability to MAX. Nothing more. This would not open up shaders for people like me, who cannot program or script shaders, but would definitely open it up for those who can (like Graphix, AmenMoses, and others who may be familiar with how shaders work or are willing to learn). Baby steps. Let's get Wicked Engine shaders available in MAX first. If someone wants to create a shader editor, then that at least is an option for them whereas it's isn't right now since we don't have any access to creating shaders in MAX at the moment.

Shaders do a lot more than just make things like glass and water effects. They can potentially make fur and hair, change materials to look like they are hand-drawn, make cell shading a possibility, enhance particle effects, do things like make snow and moss appear on the appropriate side of geometry based on a number of parameters, do things beyond what standard PBR materials can do (and more than just add animation to materials). Depending on what Wicked supports, we could potentially end up with access to both pixel shaders and vertex shaders. If there are vertex shaders, then not only could we potentially modify material properties, but also the actual model itself (such as morphing a model from one shape to another, etc.). 

Access to shaders in MAX would open up a world of professional possibilities that simply are not doable apart from them.

**Comment by earthling45** [comment]
> They can potentially make fur and hair, change materials to look like they are hand-drawn, make cell shading a possibility, enhance particle effects, do things like make snow and moss appear on the appropriate side of geometry based on a number of parameters, do things beyond what standard PBR materials can do (and more than just add animation to materials).

Than you are adding a pattern map (where should moss appear) a noise map for density of fur and of course parallax for fur to create the layers.
For moss and puddles, water or snow and ice i think that there should be support for sbsar files in Max, than you'll have access to the parameters which control those layers.
But also, let's say a wood material can be used more than once i think because the look of the material can be altered quite a lot by changing the roughness, color, grains, and so on.
I'm curious about how this will compare to loading textures, if we were to use material files.

**Comment by LeeBamberTGC** [comment]
LINKED ISSUES:
https://github.com/TheGameCreators/GameGuruRepo/issues/3142
https://github.com/TheGameCreators/GameGuruRepo/issues/3050
https://github.com/TheGameCreators/GameGuruRepo/issues/2948
https://github.com/TheGameCreators/GameGuruRepo/issues/2554
https://github.com/TheGameCreators/GameGuruRepo/issues/1176
